include/maliput_sparse/parser/validator.h

```diff
@@ -0,0 +1,115 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "maliput_sparse/parser/parser.h"
+
+namespace maliput_sparse {
+namespace parser {
+
+/// ValidatorOptions struct that contains the options for the Validator.
+struct ValidatorOptions {
+  /// Verifies adjacency for lanes in a segment.
+  bool lane_adjacency{true};
+};
+
+/// ValidatorConfig struct that contains the configuration for the Validator.
+struct ValidatorConfig {
+  double linear_tolerance{1e-12};
+};
+
+/// After parsing a road network, the Validator can be used to check for errors before
+/// creating a maliput::api::RoadNetwork via the maliput_sparse::loader::RoadNetworkLoader.
+/// The Validator can be configured to check for different types of errors and provide an interface to retrieve
+/// the errors found.
+/// The errors are stored in a vector of Error structs. The Error struct contains a message, the type of error, and the
+/// severity. It's on the user to decide how to handle the errors.
+class Validator {
+ public:
+  /// Error struct that contains the error message, type, and severity.
+  struct Error {
+    /// The type of error.
+    enum class Type {
+      kLaneAdjacency,
+    };
+    /// The severity of the error.
+    enum class Severity {
+      kWarning,
+      kError,
+    };
+
+    /// Equality operator for Error.
+    bool operator==(const Error& other) const;
+    /// Inequality operator for Error.
+    bool operator!=(const Error& other) const;
+
+    /// Message describing the error.
+    std::string message;
+    /// The type of error.
+    Type type;
+    /// The severity of the error.
+    Severity severity;
+  };
+
+  /// Constructor for Validator.
+  /// During construction, the Validator will perform the validation checks.
+  ////
+  /// @param parser The maliput_sparse::parser::Parser instance to validate.
+  /// @param options The maliput_sparse::parser::ValidatorOptions to use.
+  /// @param config The maliput_sparse::parser::ValidatorConfig to use.
+  Validator(const Parser* parser, const ValidatorOptions& options, const ValidatorConfig config);
+
+  /// Returns the errors found during validation.
+  const std::vector<Error>& GetErrors() const;
+
+ private:
+  // Helper functions for reporting errors.
+  // @param message The error message.
+  // @Param type The type of error.
+  // @Param severity The severity of the error.
+  void Report(const std::string& message, const Error::Type& type, const Error::Severity& severity);
+
+  // Method to validate lane adjacency.
+  // @param parser The maliput_sparse::parser::Parser instance to validate.
+  // @param config The maliput_sparse::parser::ValidatorConfig to use.
+  void ValidateLaneAdjacency(const Parser* parser, const ValidatorConfig config);
+
+  // Helper function to geometrically check lane adjacency.
+  void CheckAdjacency(const Lane& lane, const Lane& adjacent_lane, bool left, double tolerance);
```

Comment:
`ValidateLaneAdjacency` method goes over the different cases to validate the lane adjacency and add the error (via the private method `report`) to the list.
We can move it outside the class and make it return a vector with the errors. However, why will it be used as a standalone and not with the Validator class/functor?

Regarding `CheckAdjacency` it is just a helper function that also reports the error. We could make it return a boolean and put it outside the class(within `maliput_sparse::geometry`). Anyways, what it is doing is pretty basic: calling `maliput_sparse::geometry::ComputeDistance` and checking against the tolerance.

Reply:
> We can move it outside the class and make it return a vector with the errors. However, why will it be used as a standalone and not with the Validator class/functor?

Composability. You want to keep the validator algorithm agnostic of each validation itself. Instead, you might want to parse validation cases and run a different strategy for each of them. You can do it as you would do with yaml the encode and decode based on the error type / value to be parsed.

Reply:
I ran into this situation. Context:
1. We want to collapse Error type and Validator options together.
2. We want to differentiate LogcalLaneAdjacency errors from GeometricalLaneAdjacency errors.

The thing is that it does not make that much of sense to be able to select just geometricalLaneAdjacency validation. Why? Well, because the selection of which lanes should be checked for adjacency are determined via the graph of lanes, and for such a thing we need to have a correct structure of lanes in the segment logically validated. (Checking geometrical adjacency without checking logical adjacency first makes it impossible to check the former)

Then having different collections for errors and options like:

```cpp

struct ValidatorOptions {
  /// Verifies adjacency for lanes in a segment.
  bool lane_adjacency{true};
};
```
```cpp

class Validator {
 public:
  /// Error struct that contains the error message, type, and severity.
  struct Error {
    /// The type of error.
    enum class Type {
      kLogicalLaneAdjacency,
      kGeometricalLaneAdjacency,
    };
    ....
```

still makes sense. I am still thinking on another probable usecase for supporting this code.

Reply:
Done.